pkg/clients/postgresql/postgresql.go

```diff
@@ -3,38 +3,76 @@ package postgresql
 import (
 	"context"
 	"database/sql"
+	"fmt"
 
 	"github.com/crossplane-contrib/provider-sql/pkg/clients/xsql"
 
 	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
 	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
 )
 
+const (
+	panicNilPtr = "function call on nil pointer"
+)
+
 type postgresDB struct {
 	dsn      string
+	username string
+	password string
+	database string
 	endpoint string
 	port     string
 }
 
 // New returns a new PostgreSQL database client.
+// The default database name is an empty string.
+// The underlying pq library will default to either
+// using the value of PGDATABASE, or if unset, the
+// hardcoded db name 'postgres'.
+// The db name can be overridden later using SetDatabase,
+// which will trigger a re-format of the DSN.
 func New(creds map[string][]byte) xsql.DB {
```

Comment:
Could we also just extend the `New()` with a `database` parameter?
IMHO this would make for a more ergonomic API than using a separate `SetDatabase()` call.

Reply:
We could allow the database to be set with `New()` as well, and that would be useful to add a field to `ProviderConfig` to set the default database name - which is effectively a nicer way to set `PGDATABASE` on the `ControllerConfig`. 

However - I don't think that helps with the core use-case I'm approaching here which is at runtime, we might want to resolve the database name to something after `Connect` has already been called.

It's worth thinking about this in combination with the `Extension` draft PR (#35). To install an `Extension` into a particular database on the instance, we have to connect to postgres on that database. The database itself might be created by crossplane as well, so might not exist when `Connect` is called if it is resolved as part of a Composition.

Since we don't actually connect to the instance until we're about to run a query, we can just override the DSN prior to executing it.

I thought doing this as a separate method would make sense, rather than modifying the existing query / exec / scan methods (which would've required changing their usage in the MySQL client as well).

I suppose we could allow the database name to be set in `New()` and then override the entire `xsql.DB` instance when it changes, which would avoid the additional `SetDatabase` method, but that would require us to store the credentials data somewhere outside `xsql.DB` to allow us to recreate the instance with the correct details.

Reply:
What I don't understand is, we have all information we need in `Connect()` (and therefore when calling `New()`), no? Why would it be necessary to change the DB later on?
The `Connect()` function is called on every reconcile run and can use the information of the resource which is being reconciled (e.g. the `Extension`).

Reply:
Cripes... yeah you're right 🤦‍♂️ - not sure what I was thinking! 

We just `Connect()` with a DB name that sets the DSN, and if no 'override' DB name is set, we just continue on as normal. If we resolve a DB name, we know that value for the next time `Connect()` is called.

Back later with a fixed PR 😂